Text edit

fsih opened the pull request:
Progress toward https://github.com/LLK/scratch-paint/issues/25
Adds text editing to the text tool!

See diff from initial text tool PR here:
https://github.com/fsih/scratch-paint/compare/textTool...fsih:textEdit?expand=1

When text is in edit mode, add a text area directly on top of it. This way we get all text editing capabilities of text areas.

Some notes
- The text edit area needs to stay on top of the text appearing on the canvas, no matter what orientation it's in. If the text area is able to be scrolled or wraps, it will be out of sync. The text can also have some crazy transformations applied to it through the select tool. This math is tricky, so this should be thoroughly tested.
- Zooming in and scrolling around the canvas also changes the matrix of the text area. I needed to add state to keep track of when that happened (state.scratchPaint.viewBounds)
- Added undo state to the text tool. When the user stops typing for 1 second, a new undo state is created. This is a pretty long timeout. I made it this long because sometimes kids aren't familiar with keyboard layouts. We might want to play with this number.

fsih committed:
Make the text tool a real tool

fsih committed:
Basic text edit

fsih committed:
Mode switching

fsih committed:
Add dotted lines in text edit mode

fsih committed:
Make the guide stay in the shape of the text when text is rotated

fsih committed:
Fix some bugs

fsih committed:
Ignore nudge when text field focused

fsih committed:
Unfocus text fields when canvas clicked

fsih committed:
Unfocus text fields when canvas clicked

fsih committed:
Move the text edit target to the state and make fill work

fsih committed:
Change proptype

fsih committed:
Merge branch 'develop' into textTool

fsih committed:
Remove unnecessary changes

fsih committed:
Add a text edit area

fsih committed:
Merge branch 'develop' into textTool

fsih committed:
Showing text on top of other text

Commit:
Edit text moves with text transform

fsih committed:
Add undo

fsih committed:
Add transform when zoomed in

fsih committed:
Detect view bounds changes and update the edit text matrix in response

fsih committed:
Check for text area in a better way

fsih committed:
fix lint

fsih committed:
Add missing file

fsih committed:
lint

fsih committed:
Don't apply outline to text by default

fsih committed:
Make clicking outside of the canvas activate select mode. This way the user can change the fill color easily, instead of the text getting deselected as soon as they click fill.

fsih commented:
I made the text tool draw no outline by default, as recommended by Paul

paulkaplan commented:
Just a few things I noticed playing with this. I'll probably create separate issues after this is merged, but wanted to have them down here for posterity. 

When clicking another text area while editing a text area, the currently editing text area should both shift out of editing mode (which it does) _and_ shift out of select mode (which it does not).
![text-select-1](https://user-images.githubusercontent.com/654102/37912253-17c280a0-30e0-11e8-93ee-e4a752a0c2c3.gif)

When in text editing mode, after exiting a text area (so that it is selected but not editing), you cannot re-enter editing on that text area without clicking outside to deselect it first
![text-select-2](https://user-images.githubusercontent.com/654102/37912295-3ae567c8-30e0-11e8-9c13-801785affedd.gif)

When in the normal select mode, there is a heavy expectation that double clicking on text would enter editing mode (from most GUIs with text areas, including scratch 2), but it doesn't. 
![text-select-3](https://user-images.githubusercontent.com/654102/37912391-6e90ea20-30e0-11e8-9fac-c48ee01c90c8.gif)

When in normal select mode, with a text area selected, clicking the text mode tool in scratch 2 transitions that text area into editing mode. Here it simply deselects everything. In the case where 
 multiple things are selected, scratch 2 deselects everything. 
![text-select-4](https://user-images.githubusercontent.com/654102/37912564-ca4b9734-30e0-11e8-813a-00b1f7ca054f.gif)

paulkaplan committed:
Fix double clicking to enter edit mode and clearing selection.

The first was caused by the "lastMode" flag being stored after the
double click edit was done, causing it to immediately exit edit mode.
The second problem was that the selection should be cleared when
entering edit mode on another text area.

paulkaplan commented:
Fixed the first and second issue with a commit: af2c49e

paulkaplan commented on src/helper/tools/text-tool.js:
```diff
@@ -0,0 +1,326 @@
+            document.body.style.cursor = 'auto';
+        }
+    }
+    handleMouseDown (event) {
```
There is some seriously complicated logic in this function. I'd like for us to come back to this and break it into more clear state transitions. It took me ~20 minutes to understand the entire FROM => TO transitions!

paulkaplan reviewed (approved):
This LG for now. There are obviously some follow ups we'll need to make (a few were listed in the comments) but this feels good enough to start testing with.
